transports/webrtc/src/webrtc_connection.rs

```diff
@@ -0,0 +1,295 @@
+
+        // 1. OFFER
+        let offer = peer_connection.create_offer(None).await?;
+        log::debug!("OFFER: {:?}", offer.sdp);
+        peer_connection.set_local_description(offer).await?;
+
+        // 2. ANSWER
+        // Set the remote description to the predefined SDP.
+        let remote_ufrag = remote_fingerprint.to_ufrag();
+        let server_session_description = render_description(
+            sdp::SERVER_SESSION_DESCRIPTION,
+            addr,
+            remote_fingerprint,
+            &remote_ufrag,
+        );
+        log::debug!("ANSWER: {:?}", server_session_description);
+        let sdp = RTCSessionDescription::answer(server_session_description).unwrap();
+        // NOTE: this will start the gathering of ICE candidates
+        peer_connection.set_remote_description(sdp).await?;
+
+        return Ok(Self { peer_connection });
+    }
+
+    pub async fn accept(
+        addr: SocketAddr,
+        config: RTCConfiguration,
+        udp_mux: Arc<dyn UDPMux + Send + Sync>,
+        our_fingerprint: &Fingerprint,
+        remote_ufrag: &str,
+    ) -> Result<Self, Error> {
+        // Set both ICE user and password to our fingerprint because that's what the client is
+        // expecting (see [`Self::connect`] "2. ANSWER").
+        let ufrag = our_fingerprint.to_ufrag();
+        let mut se = Self::setting_engine(udp_mux, &ufrag, addr.is_ipv4());
+        {
+            se.set_lite(true);
+            se.disable_certificate_fingerprint_verification(true);
+            // Act as a DTLS server (one which waits for a connection).
+            //
+            // NOTE: removing this seems to break DTLS setup (both sides send `ClientHello` messages,
+            // but none end up responding).
+            se.set_answering_dtls_role(DTLSRole::Server)?;
+        }
+
+        let api = APIBuilder::new().with_setting_engine(se).build();
+        let peer_connection = api.new_peer_connection(config).await?;
+
+        let client_session_description = render_description(
+            sdp::CLIENT_SESSION_DESCRIPTION,
+            addr,
+            &Fingerprint::new_sha256("NONE".to_owned()), // certificate verification is disabled, so any value is okay.
+            remote_ufrag,
+        );
+        log::debug!("OFFER: {:?}", client_session_description);
+        let sdp = RTCSessionDescription::offer(client_session_description).unwrap();
+        peer_connection.set_remote_description(sdp).await?;
+
+        let answer = peer_connection.create_answer(None).await?;
+        // Set the local description and start UDP listeners
+        // Note: this will start the gathering of ICE candidates
+        log::debug!("ANSWER: {:?}", answer.sdp);
+        peer_connection.set_local_description(answer).await?;
+
+        return Ok(Self { peer_connection });
+    }
+
+    pub async fn create_initial_upgrade_data_channel(
+        &self,
+        negotiated: Option<bool>,
```

Comment:
https://github.com/libp2p/rust-libp2p/pull/2622#discussion_r918934992

so basically because we create initial data channel on both sides, we need to prevent one (client) from signalling a command to open a data channel to another (server). This is what Option<true> does.

Reply:
> [#2622 (comment)](https://github.com/libp2p/rust-libp2p/pull/2622#discussion_r918934992)
> 
> so basically because we create initial data channel on both sides, we need to prevent one (server) from signalling a command to open a data channel to another (client). This is what Option does.

Okay, thanks for the explanation! So what we are basically expressing here is the role the node will have in the channel? Would it make sense to use https://docs.rs/libp2p/0.46.1/libp2p/core/enum.Endpoint.html here in the public API and map that to the necessary webrtc field?

I am asking because `Option<bool>` has 3 states:

- `None`
- `Some(true)`
- `Some(false)`

And if we are really expressing the dialer/listener role here then I am wondering what the 3rd state will do? In any case, the way it is modeled is an (IMO ugly) implementation detail of the webrtc library and seems worthwhile hiding.

Reply:
changed it to `bool`. does it look better now?

"in the public API" how's this a public API if it's only exposed inside a crate?

Reply:
> changed it to `bool`. does it look better now?

It is better and we can leave it at that if you want :)
I am not exactly a friend of boolean-type parameters because it is not clear what the flag does without reading documentation and or the code. This is also known as a "boolean trap".

> "in the public API" how's this a public API if it's only exposed inside a crate?

It is the public API of the type. Just because a type isn't exposed on the public API of a crate doesn't mean we shouldn't apply principles like information hiding in their API design. It makes the code much easier to understand if a bigger module is grouped into smaller, coherent modules. That is why I am pushing for newtypes and encapsulation. It allows the code to express through it structure, which components belong together, which invariants have to hold for what functionality etc :)